feat(curriculum): add smart campus workshop to JS cert (#69120)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/blocks/workshop-smart-campus-directory/6a7050a3eea4b3ab9ebfb889.md

---
id: 6a7050a3eea4b3ab9ebfb889
title: Step 2
challengeType: 1
dashedName: step-2
---

# --description--

In this step, you will map out the structure you're working with. 
In `rawData`, there are multiple layer of data are nested together,
like this: `department -> programs -> instructors`.

Now imagine you needed to find an instructor by their email address. With `rawData` in its 
current shape, you'd have to loop through every department, then every program, then every instructor, 
just to check one email. This is slow and repetitive — and it's exactly the problem normalization solves.

Now, you should create a function `getInstructorByEmail` using loops on object, which will take the a 
single argument email and retunn the name of the instructor.

> 

Author: Hi @jdwilkin4, 
I have an idea that if I bring the step-6 (teaching the object.entries()) step before step-2 and teach the looping over objects , then instruct camper to create a function `getInstructorByEmail` using loops like below, then campar will be able to feel the problem of nested data and their high complexity : 

```js
function getInstructorByEmail(email) {
  // Loop through each department
  for (const dept in rawData) {
    const programs = rawData[dept].programs;

    // Loop through each program in the department
    for (const program of programs) {
      const instructors = program.instructors;

      // Loop through each instructor in the program
      for (const instructor of instructors) {
        if (instructor.email === email) {
          return instructor.name; // Found the match
        }
      }
    }
  }
  return null; // If no match found
}

// Example usage:
console.log(getInstructorByEmail("fatima@edu.com"));   // "Fatima"
console.log(getInstructorByEmail("alejandro@edu.com")); // "Alejandro"
console.log(getInstructorByEmail("sofia@edu.com"));     // "Sofia"

```

Then steps for the main normilizer function can be placed after teaching above consepts.

Reviewer: This could work if broken up into steps and testing out the function along the way. Otherwise, if you have beginners write all of this code and then test it later, they will probably struggle with it. 

Here is an example breakdown of possible steps

````js
// Declare function and return null step
function getInstructorByEmail(email) {
  return null;
}

// add console.log step
function getInstructorByEmail(email) {
  return null;
}

console.log(getInstructorByEmail("fatima@edu.com"));

// loop through departments and log it to console step

function getInstructorByEmail(email) {
  for (const dept in rawData) {
    console.log(dept);
  }
  return null;
}

console.log(getInstructorByEmail("fatima@edu.com"));

// access programs step

function getInstructorByEmail(email) {
  for (const dept in rawData) {
    const programs = rawData[dept].programs;
  }
  return null;
}

console.log(getInstructorByEmail("fatima@edu.com"));

// loop through programs and log it to console step

function getInstructorByEmail(email) {
  for (const dept in rawData) {
    const programs = rawData[dept].programs;

    for (const program of programs) {
      console.log(program);
    }
  }
  return null;
}

console.log(getInstructorByEmail("fatima@edu.com"));

// etc.
````

Author: I have broken it into steps; can you please check the chronology form step-2 to step-9

---
id: 6a7050a3eea4b3ab9ebfb889
title: Step 2
challengeType: 1
dashedName: step-2
---

# --description--

In this step, you will map out the structure you're working with. 
In `rawData`, there are multiple layer of data are nested together,

> like this: `department -> programs -> instructors`.

Reviewer: ```suggestion
In `rawData`, multiple layers of data are nested within each other, following a hierarchy like this: `department → programs → instructors`.
```

Now imagine you needed to find an instructor by their email address. With `rawData` in its 
current shape, you'd have to loop through every department, then every program, then every instructor, 
just to check one email. This is slow and repetitive — and it's exactly the problem normalization solves.

Now, in this and next couple steps, you should create a function `getInstructorByEmail` using loops on object,

> which will take the a single argument email and retunn the name of the instructor.

Reviewer: The current phrasing isn't clear on the true objective. Right now it reads as loops are inefficient but we are going to use them anyway. The wording should be updated to say we are using loops to show the issue and then see how normalization solves it. 

This is clearer wording.

```suggestion
At this point, `rawData` is deeply nested: `departments → programs → instructors`. To find an instructor by email, you would have to loop through each department, then each program, and finally each instructor until you find a match.

While that approach works, it becomes slow and repetitive as the data grows. 

Later on in the workshop, you will learn how to normalize the data which is a more efficient approach. 

But for the next few steps, you will create a function that takes an email as its argument, use loops to search through the nested `rawData` structure, and return the instructor's name when it finds a matching email address.

```

> To brgin with, declare a function `getInstructorByEmail` and return `null` from the function.

Reviewer: ```suggestion
Declare a function called `getInstructorByEmail` and use `email` for the parameter name. Then return `null` from the function.
```

---
id: 6a7050a3eea4b3ab9ebfb889
title: Step 2
challengeType: 1
dashedName: step-2
---

# --description--

In this step, you will map out the structure you're working with. 

> In `rawData`, multiple layers of data are nested within each other, following a hierarchy like this: `department → programs → instructors`.

Reviewer: This whole description is now to long. Remove this part. 

```suggestion
```

At this point, `rawData` is deeply nested: `departments → programs → instructors`. To find an instructor by email, you would have to loop through each department, then each program, and finally each instructor until you find a match.

While that approach works, it becomes slow and repetitive as the data grows. 

Later on in the workshop, you will learn how to normalize the data which is a more efficient approach. 

But for the next few steps, you will create a function that takes an email as its argument, use loops to search through the nested `rawData` structure, and return the instructor's name when it finds a matching email address.

> 

Reviewer: remove extra line

```suggestion
```
